massa-consensus-worker/src/state/mod.rs

```diff
@@ -199,125 +215,150 @@ impl ConsensusState {
+    ///
+    /// do the following 3 times:
+    ///      extend kept_blocks with the parents of the current kept_blocks
+    ///      fill holes by adding to kept_blocks all the active block IDs whose slot is after the earliest kept_blocks of their thread (included) (but before end_slot (included) if it is Some)
+    ///
+    /// return kept_blocks
+    pub fn list_required_active_blocks(
+        &self,
+        end_slot: Option<Slot>,
+    ) -> Result<PreHashSet<BlockId>, ConsensusError> {
+        // if an end_slot is provided compute the lastest final block for that given slot
+        // if not use the latest_final_blocks_periods
+        let effective_latest_finals: Vec<(BlockId, u64)> = if let Some(slot) = end_slot {
+            self.list_latest_final_blocks_at(slot)?
+        } else {
+            self.latest_final_blocks_periods.clone()
+        };
 
-            for retain_h in retain_clone.into_iter() {
-                retain_active.extend(
-                    self.get_full_active_block(&retain_h)
-                        .ok_or_else(|| ConsensusError::ContainerInconsistency(format!("inconsistency inside block statuses pruning and retaining the parents of the selected blocks - {} is missing", retain_h)))?
-                        .0.parents
-                        .iter()
-                        .map(|(b_id, _p)| *b_id),
-                )
-            }
+        // init kept_blocks using effective_latest_finals
+        let mut kept_blocks: PreHashSet<BlockId> = effective_latest_finals
+            .iter()
+            .map(|(id, _period)| *id)
+            .collect();
 
-            // find earliest kept slots in each thread
-            let mut earliest_retained_periods: Vec<u64> = self
-                .latest_final_blocks_periods
-                .iter()
-                .map(|(_, p)| *p)
-                .collect();
-            for retain_h in retain_active.iter() {
-                let retain_slot = &self
-                    .get_full_active_block(retain_h)
-                    .ok_or_else(|| ConsensusError::ContainerInconsistency(format!("inconsistency inside block statuses pruning and finding earliest kept slots in each thread - {} is missing", retain_h)))?
-                    .0.slot;
-                earliest_retained_periods[retain_slot.thread as usize] = std::cmp::min(
-                    earliest_retained_periods[retain_slot.thread as usize],
-                    retain_slot.period,
-                );
+        // add all the active blocks that are after the effective_latest_finals of their thread
+        self.add_active_blocks_after(&mut kept_blocks, &effective_latest_finals, end_slot);
+
+        // do the following 3 times
+        for _ in 0..3 {
+            // extend kept_blocks with the parents of the current kept_blocks
+            let kept_blocks_clone = kept_blocks.clone();
```

Comment:
kept_blocks is potentially large. Instead of cloning it, it might be better to iterate over the current kept_blocks, accumulate the new items to be added in a separate container, and then add them to kept_blocks at the end of the outer loop iteration